usr.sbin/rmt/rmt.8
.\"	$NetBSD: rmt.8,v 1.13.40.1 2009/05/13 19:20:37 jym Exp $
.\"
.\" Copyright (c) 1983, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)rmt.8	8.3 (Berkeley) 6/1/94
.\"
.Dd June 1, 1994
.Dt RMT 8
.Os
.Sh NAME
.Nm rmt
.Nd remote magtape protocol module
.Sh SYNOPSIS
.Nm
.Sh DESCRIPTION
.Nm
is a program used by the remote dump and restore programs
in manipulating a magnetic tape drive through an interprocess
communication connection.
.Nm
is normally started up with an
.Xr rexec 3
or
.Xr rcmd 3
call.
.Pp
The
.Nm
program accepts requests specific to the manipulation of
magnetic tapes, performs the commands, then responds with
a status indication.  All responses are in
.Tn ASCII
and in
one of two forms.
Successful commands have responses of:
.Bd -filled -offset indent
.Sm off
.Sy A Ar number No \en
.Sm on
.Ed
.Pp
.Ar Number
is an
.Tn ASCII
representation of a decimal number.
Unsuccessful commands are responded to with:
.Bd -filled -offset indent
.Sm off
.Xo Sy E Ar error-number
.No \en Ar error-message
.No \en
.Xc
.Sm on
.Ed
.Pp
.Ar Error-number
is one of the possible error
numbers described in
.Xr intro 2
and
.Ar error-message
is the corresponding error string as printed
from a call to
.Xr perror 3 .
The protocol comprises the
following commands, which are sent as indicated - no spaces are supplied
between the command and its arguments, or between its arguments, and
.Ql \en
indicates that a newline should be supplied:
.Bl -tag -width Ds
.It Sy \&O Ns Ar device Ns \en Ns Ar mode Ns \en
Open the specified
.Ar device
using the indicated
.Ar mode .
.Ar Device
is a full pathname and
.Ar mode
is an
.Tn ASCII
representation of a decimal
number suitable for passing to
.Xr open 2 .
If a device had already been opened, it is
closed before a new open is performed.
.It Sy C Ns Ar device Ns \en
Close the currently open device.  The
.Ar device
specified is ignored.
.It Sy L Ns Ar offset Ns \en Ns Ar whence Ns \en
Perform an
.Xr lseek 2
operation using the specified parameters.
The response value is that returned from the
.Xr lseek 2
call.
.Sm off
.It Sy W Ar count No \en
.Sm on
Write data onto the open device.
.Nm
reads
.Ar count
bytes from the connection, aborting if
a premature end-of-file is encountered.
The response value is that returned from
the
.Xr write 2
call.
.Sm off
.It Sy R Ar count No \en
.Sm on
Read
.Ar count
bytes of data from the open device.
If
.Ar count
exceeds the size of the data buffer (10 kilobytes), it is
truncated to the data buffer size.
.Nm
then performs the requested
.Xr read 2
and responds with
.Sm off
.Sy A Ar count-read No \en
.Sm on
if the read was
successful; otherwise an error in the
standard format is returned.  If the read
was successful, the data read is then sent.
.It Sy I Ns Ar operation Ns \en Ns Ar count Ns \en
.Sm on
Perform a
.Dv MTIOCOP
.Xr ioctl 2
command using the specified parameters.
The parameters are interpreted as the
.Tn ASCII
representations of the decimal values
to place in the
.Ar mt_op
and
.Ar mt_count
fields of the structure used in the
.Xr ioctl 2
call.  The return value is the
.Ar count
parameter when the operation is successful.
.It Sy S
Return the status of the open device, as
obtained with a
.Dv MTIOCGET
.Xr ioctl 2
call.  If the operation was successful,
an ``ack'' is sent with the size of the
status buffer, then the status buffer is
sent (in binary).
.El
.Pp
Any other command causes
.Nm
to exit.
.Sh DIAGNOSTICS
All responses are of the form described above.
.Sh SEE ALSO
.Xr rcmd 3 ,
.Xr rexec 3 ,
.Xr mtio 4 ,
.Xr rdump 8 ,
.Xr rrestore 8
.Sh HISTORY
The
.Nm
command appeared in
.Bx 4.2 .
.Sh BUGS
People should be discouraged from using this for a remote
file access protocol.


doc/CHANGES
# LIST OF CHANGES FROM LAST RELEASE:			<$Revision: 1.1231 $>
#
#
# [Note: This file does not mention every change made to the NetBSD source tree.
# For an exhaustive list, please subscribe to the `source-changes' mailing list,
# or see the mailing list archives.  For more information, send a message
# containing just the word `help' to `majordomo@NetBSD.org'.]
#
# ----------------------------------------------------------------------------
# PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:
#
#	theme: Content ...
#		more content ...
#		more content. [developer DATE-FORMAT]
#
# That's it, one tab to start an item, and two tabs to indent the content.
# Also please start the content with a capital and end it with a dot.
#
# This is very important for the www scripts, otherwise it won't be able to
# parse the file and the HTML output will be mangled.
#
# See htutils/changes/changes2html script for more details.
#
# ----------------------------------------------------------------------------


Changes from NetBSD 5.0 to NetBSD 6.0:

	syslog(3): Import Martin Schuette's Summer of Code project: 
		Reliable TCP connections, syslog protocol API,
		encryption. [christos 20081031]
	agp(4): Add support for Intel G35. [christos 20081108]
	kernel: Add generic producer/consumer queue. [matt 20081111]
	lkm: Removed. Superseded by the new module framework. [ad 20081112]
	module: Allow the POSIX semaphore code to be loaded as a module.
		[ad 20081112]
	module: Make compat_ossaudio a module. [ad 20081113]
	module: Make compat_ndis a module. [ad 20081113]
	build.sh: Add support for new MAKEVERBOSE levels 3 and 4.
		[apb 20081113]
	libfetch: Update to libfetch 2.20. Allow HTTP auth to be provided in
		URL. [joerg 20081114]
	module: Make POSIX AIO loadable as a module. [ad 20081114]
	amd64, i386: Make boot loader autoload a kernel module corresponding
		to the root file system type. [ad 20081119]
	module: Make the emulations, exec formats, coredump, NFS, and
		the NFS server into modules. [ad 20081119]
	rump(3): Move VFS functionality away from librump to a new library:
		librumpvfs [pooka 20081119]
	kernel, libc: Add new ptree (Patricia / RADIX tree) implementation.
		[matt 20081120]
	i386: Switch the i386 installation CDs away from RAM disks and
		instead use the CD-ROM as the root file system. [ad 20081121]
	i386: Remove most compat and file system options from GENERIC
		since they are also provided by modules. [ad 20081122]
	nfs: Avoid noisy nfs_timer/nfs_reply DEBUG output that occurs
		when the NFS server goes away. [mrg 20081123]
	compat_osf1(8): Add getpgid, needed for Tru64 4.0g ksh.
		[njoly 20081124]
	module: Make PPP compressors modules. [cube 20081125]
	i386: Make pxeboot(8) understand boot.cfg(5). [ad 20081125]
	agp(4): Add support for the Intel G45 AGP. [christos 20081129]
	compat_osf1(8): Fix getdirentries(2). [njoly 20081202]
	module: Add a couple of sysctl(8) entries, kern.module.autoload and
		kern.module.verbose. [ad 20081203]
	module: Add a layerfs module. [ad 20081205]
	libc: Add fdopendir(3), from Solaris/Linux. [ad 20081205]
	compat_linux32: Add getresuid() and getresgid() syscalls.
		[njoly 20081205]
	ddb(4): Add 'show vmem' and 'show all vmems' commands.
		[cegger 20081207]
	re(4): Add support for a 8168C/8111C revision. [alc 20081208]
	tab(1): Add tabs utility, fixes PR bin/11227. [roy 20081211]
	httpd(8): Add rc.d script to control httpd daemon. [mishka 20081211]
	ath(4): Switch from the binary HAL to the open-source HAL released by
		Sam Leffler. [alc 20081211]
	compat_linux32: Add clock_nanosleep syscall. [njoly 20081212]
	boot.cfg(5): Add an option to clear the screen. [christos 20081213]
	pms(4): Add Elantech multi-touch trackpad support [jmcneill 20081213]
	sparc: Switch to use wscons. [macallan 20081213]
	snprintb(3), kprintf(9): Share snprintb(3) among kernel and libutil
		and make it supersede bitmask_snprintf(9). [christos 20081216]
	sup(1): Add -M user flag to specify the username to send mail to from
		the command line. [christos 20081217]
	malloc(9): Remove MALLOC() and FREE() macros. [cegger 20081217]
	x68k:	pow(4) switches to use MI powerd(8) [isaki 20081220]
	viaide(4): Add VT8237S Integrated SATA Controller support.
		[nonaka 20081221]
	lvm:	Add LVM functionality to NetBSD; this is done by dm(4) driver
		in kernel and lvm2 tools made by RedHat. We use the same well
		known interface as Linux does for managing LVM. [haad 20081222]
	gcscaudio(4): Add AMD Geode CS5536 audio driver, from SHIMIZU Ryo.
		[jmcneill 20081228]
	mount_cd9660(8): Add nocasetrans option as a synonym for nomaplcase.
		[christos 20090103]
	rc.d(8): Add a script for rndctl(8). [apb 20090104]
	module: Add sysvbfs module. [pooka 20090105]
	bwi(4): Add a driver for Broadcom AirForce / Apple Airport Extreme
		wireless network controllers. From Taylor Campbell.
		[macallan 20090109]
	rgephy(4): Support for RTL8211C(L) phy. From FreeBSD. [cegger 20090109]
	rgephy(4): Support for flow control. From OpenBSD. [cegger 20090109]
	makefs(8): Add support for encoding RISC OS metadata into
		ISO 9660 file systems. [bjh21 20090110]
	time_t, dev_t: Change time_t and dev_t to be 64 bit quantities. Fixed
		POSIX issues in timeval and timespec. [christos 20090110]
	make(1): Changes to the way .for loops are processed. [dsl 20090111]
	openpgpsdk: Initial import of openpgpsdk v0.9. [agc 20090112]
	age(4):	Add a driver for Attensic/Atheros L1 Gigabit Ethernet.
		From FreeBSD via OpenBSD. [cegger 20090116]
	atphy(4): Add a driver for Attensic/Atheros F1 PHY.
		From FreeBSD via OpenBSD. [cegger 20090116]
	atf(7): Import 0.6.  [jmmv 20090119]
	msdosfs: The default behavior when mounting empty file systems has
		been changed. Previously, msdosfs would assume the -s flag
		(short filenames) in this case; now msdosfs will assume the
		-l flag (long filenames). [jmcneill 20090122]
	openpgpsdk: Imported openpgpsdk 0.9 library (BSD-licensed) into
		src/crypto/external. The command-line driver program, openpgp,

usr.sbin/rmt/rmt.c
/*	$NetBSD: rmt.c,v 1.15.6.1 2009/05/13 19:20:37 jym Exp $	*/

/*
 * Copyright (c) 1983, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#ifndef lint
__COPYRIGHT("@(#) Copyright (c) 1983, 1993\
 The Regents of the University of California.  All rights reserved.");
#endif /* not lint */

#ifndef lint
#if 0
static char sccsid[] = "@(#)rmt.c	8.1 (Berkeley) 6/6/93";
#else
__RCSID("$NetBSD: rmt.c,v 1.15.6.1 2009/05/13 19:20:37 jym Exp $");
#endif
#endif /* not lint */

/*
 * rmt
 */
#include <sys/types.h>
#include <sys/ioctl.h>
#include <sys/mtio.h>
#include <sys/socket.h>
#include <sys/stat.h>

#include <errno.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

int	tape = -1;

int	maxrecsize = -1;

#define	SSIZE	64
char	device[SSIZE];
char	count[SSIZE], mode[SSIZE], pos[SSIZE], op[SSIZE];

char	resp[BUFSIZ];

FILE	*debug;
#define	DEBUG(f)	if (debug) fprintf(debug, f)
#define	DEBUG1(f,a)	if (debug) fprintf(debug, f, a)
#define	DEBUG2(f,a1,a2)	if (debug) fprintf(debug, f, a1, a2)

char	*checkbuf __P((char *, int));
void	 error __P((int));
int	 main __P((int, char **));
void	 getstring __P((char *, size_t));

int
main(argc, argv)
	int argc;
	char **argv;
{
	char *record = NULL;
	int rval;
	char c;
	int n, i, cc;

	argc--, argv++;
	if (argc > 0) {
		debug = fopen(*argv, "w");
		if (debug == 0)
			exit(1);
		(void)setbuf(debug, (char *)0);
	}
top:
	errno = 0;
	rval = 0;
	if (read(STDIN_FILENO, &c, 1) != 1)
		exit(0);
	switch (c) {

	case 'O':
		if (tape >= 0)
			(void) close(tape);
		getstring(device, sizeof(device));
		getstring(mode, sizeof(mode));
		DEBUG2("rmtd: O %s %s\n", device, mode);
		tape = open(device, atoi(mode),
		    S_IRUSR|S_IWUSR|S_IRGRP|S_IWGRP|S_IROTH|S_IWOTH);
		if (tape < 0)
			goto ioerror;
		goto respond;

	case 'C':
		DEBUG("rmtd: C\n");
		getstring(device, sizeof(device));	/* discard */
		if (close(tape) < 0)
			goto ioerror;
		tape = -1;
		goto respond;

	case 'L':
		getstring(count, sizeof(count));
		getstring(pos, sizeof(pos));
		DEBUG2("rmtd: L %s %s\n", count, pos);
		rval = lseek(tape, (off_t)strtoll(count, NULL, 10), atoi(pos));
		if (rval < 0)
			goto ioerror;
		goto respond;

	case 'W':
		getstring(count, sizeof(count));
		n = atoi(count);
		DEBUG1("rmtd: W %s\n", count);
		record = checkbuf(record, n);
		for (i = 0; i < n; i += cc) {
			cc = read(STDIN_FILENO, &record[i], n - i);
			if (cc <= 0) {
				DEBUG("rmtd: premature eof\n");
				exit(2);
			}
		}
		rval = write(tape, record, n);
		if (rval < 0)
			goto ioerror;
		goto respond;

	case 'R':
		getstring(count, sizeof(count));
		DEBUG1("rmtd: R %s\n", count);
		n = atoi(count);
		record = checkbuf(record, n);
		rval = read(tape, record, n);
		if (rval < 0)
			goto ioerror;
		(void)snprintf(resp, sizeof(resp), "A%d\n", rval);
		(void)write(STDOUT_FILENO, resp, strlen(resp));
		(void)write(STDOUT_FILENO, record, rval);
		goto top;

	case 'I':
		getstring(op, sizeof(op));
		getstring(count, sizeof(count));
		DEBUG2("rmtd: I %s %s\n", op, count);
		{
			struct mtop mtop;

			mtop.mt_op = atoi(op);
			mtop.mt_count = atoi(count);
			if (ioctl(tape, MTIOCTOP, (char *)&mtop) < 0)
				goto ioerror;
			rval = mtop.mt_count;
		}
		goto respond;

	case 'S':		/* status */
		DEBUG("rmtd: S\n");
		{
			struct mtget mtget;

			if (ioctl(tape, MTIOCGET, (char *)&mtget) < 0)
				goto ioerror;
			rval = sizeof (mtget);
			/* limit size to 'original' mtget size */
			if (rval > 24)
				rval = 24;
			(void)snprintf(resp, sizeof(resp), "A%d\n", rval);
			(void)write(STDOUT_FILENO, resp, strlen(resp));
			(void)write(STDOUT_FILENO, (char *)&mtget, rval);
			goto top;
		}

	default:
		DEBUG1("rmtd: garbage command %c\n", c);
		exit(3);
	}
respond:
	DEBUG1("rmtd: A %d\n", rval);
	(void)snprintf(resp, sizeof(resp), "A%d\n", rval);
	(void)write(STDOUT_FILENO, resp, strlen(resp));
	goto top;
ioerror:
	error(errno);
	goto top;
}

void
getstring(bp, size)
	char *bp;
	size_t size;
{
	char *cp = bp;
	char *ep = bp + size - 1;

	do {
		if (read(STDIN_FILENO, cp, 1) != 1)
			exit(0);
	} while (*cp != '\n' && ++cp < ep);
	*cp = '\0';
}

char *
checkbuf(record, size)
	char *record;
	int size;
{

	if (size <= maxrecsize)
		return (record);
	if (record != 0)
		free(record);
	record = malloc(size);
	if (record == 0) {
		DEBUG("rmtd: cannot allocate buffer space\n");
		exit(4);
	}
	maxrecsize = size;
	while (size > 1024 &&
	    setsockopt(0, SOL_SOCKET, SO_RCVBUF, &size, sizeof (size)) < 0)
		size -= 1024;
	return (record);
}

void
error(num)
	int num;
{

	DEBUG2("rmtd: E %d (%s)\n", num, strerror(num));
	(void)snprintf(resp, sizeof(resp), "E%d\n%s\n", num, strerror(num));
	(void)write(STDOUT_FILENO, resp, strlen(resp));
}
